### PR TITLE
feat: use functions for first and recover sets

### LINKED ISSUE #7620: Parser: Replace FIRST token sets by a function

### LINKED ISSUE #7621: Parser: Replace RECOVER sets by a function

### PR DESCRIPTION
**Pull request by herluf-ba**
Closes #7620
Closes #7621

### CODE REVIEW

**Comment by magnus-madsen** [comment]
Any performance impact?

File: main/src/ca/uwaterloo/flix/language/ast/TokenKind.scala [hunk 1]
```diff
@@ -195,6 +195,256 @@ sealed trait TokenKind {
...
+    */
+  def isCommentNonDoc: Boolean = this match {
+    case TokenKind.CommentLine | TokenKind.CommentBlock => true
+    case _ => false
+  }
+
+  /**
+    * Checks if this token is a doc, line or block comment.
+    */
+  def isComment: Boolean = this == TokenKind.CommentDoc || this.isCommentNonDoc
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/TokenKind.scala:210]: The equality check here is costly. Remember it goes into equals which probably checks that the strings are the same!

**Comment by magnus-madsen** [comment]
!!!jar

File: main/src/ca/uwaterloo/flix/language/ast/TokenKind.scala [hunk 2]
```diff
@@ -195,6 +195,256 @@ sealed trait TokenKind {
...
+  def isRecoverDecl: Boolean = this.isFirstDecl
+
+  /**
+    * Checks if kind is a TokenKind that warrants breaking an expression parsing loop.
+    * For instance, if we find an 'trait' keyword in the middle of an expression,
+    * we can assume that there is no more expression to parse.
+    * Because isRecoverExpr returns true for KeywordTrait the expression loop stops,
+    * making the parse return back into the declaration loop which can capture the trait.
+    */
+  def isRecoverExpr: Boolean = this.isFirstDecl || this == TokenKind.Semi
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/TokenKind.scala:415]: Again, this check is expensive.

**Comment by magnus-madsen** [comment]
Looks good for merge, once the uses of `==` are fixed.

File: main/src/ca/uwaterloo/flix/language/ast/TokenKind.scala (hunk 1 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/TokenKind.scala:210]: I stand corrected. These are case objects without any data, so should be fine.

**Review by magnus-madsen — APPROVED** [review]

File: main/src/ca/uwaterloo/flix/language/ast/TokenKind.scala (hunk 1 above)
Comment (herluf-ba) [reply main/src/ca/uwaterloo/flix/language/ast/TokenKind.scala:210]: `this` in this case is a `TokenKind`, which caries no data. I'll change it to all pattern-matches just to be safe though 👍

**Pull request merged**